src/rust/ensogl/src/display/style/sheets.rs

```diff
@@ -0,0 +1,1063 @@
+    pub fn apply_change(&mut self, change:Change) -> HashSet::<Index<Query>> {
+        self.apply_changes(iter::once(change))
+    }
+
+    /// Apply a set of `Change`s. Returns indexes of all affected variables.
+    pub fn apply_changes<I>(&mut self, changes:I) -> HashSet::<Index<Query>>
+    where I:IntoIterator<Item=Change> {
+        let mut changed          = HashSet::<Index<Query>>::new();
+        let mut possible_orphans = Vec::<Index<Sheet>>::new();
+        let sheets_iter = changes.into_iter().map(|change| {
+            let sheet_id = self.sheet(change.path);
+            let sheet    = &mut self.sheets[sheet_id];
+
+            // Remove expression bindings.
+            let opt_expr = mem::take(&mut sheet.expr);
+            if let Some(expr) = opt_expr {
+                for query_id in expr.args {
+                    self.queries[query_id].usages.remove(&sheet_id);
+                    self.drop_query_if_unused(query_id);
+                }
+            }
+
+            // Set new value and rebind variables.
+            let sheet = &mut self.sheets[sheet_id];
+            match change.value {
+                None => {
+                    let needs_rebind = sheet.value.is_some();
+                    if needs_rebind {
+                        sheet.value = None;
+                        for query_id in sheet.bindings.clone() {
+                            if self.rebind_query(query_id) {
+                                changed.insert(query_id);
+                            }
+                        }
+                        possible_orphans.push(sheet_id);
+                    }
+                },
+                Some(value) => {
+                    let needs_rebind = sheet.value.is_none();
+                    match value {
+                        Value::Data(data) => sheet.value = Some(data),
+                        Value::Expression(expr) => {
+                            let queries = expr.args.iter().map(|path| self.unmanaged_query(path)).collect_vec();
+                            for query_id in &queries {
+                                self.queries[*query_id].usages.insert(sheet_id);
+                            }
+                            let bound_expr = BoundExpression::new(queries,expr.function);
+                            let sheet      = &mut self.sheets[sheet_id];
+                            sheet.expr     = Some(bound_expr);
+                            self.recompute(sheet_id);
+                        }
+                    }
+                    if needs_rebind {
+                        let sheet = &self.sheets[sheet_id];
+                        for query_id in sheet.matches.clone() {
+                            if self.rebind_query(query_id) {
+                                changed.insert(query_id);
+                            }
+                        }
+                    }
+                }
+            };
+            sheet_id
+        });
```

Comment:
Was that function so huge before? Maybe it is good candidate to split for lesser functions.

Reply:
Ill try, but no rpomises. This is a very coupled function andwhen refactoring you get a lot of mess with mutable borrows. Lets see what we can do here :)